src/engagement_db_coda_sync/lib.py

```diff
@@ -1,15 +1,102 @@
+import json
+
+from coda_v2_python_client.firebase_client_wrapper import CodaV2Client
 from core_data_modules.cleaners import Codes
 from core_data_modules.cleaners.cleaning_utils import CleaningUtils
 from core_data_modules.data_models import Message as CodaMessage, Label, Origin
 from core_data_modules.logging import Logger
 from core_data_modules.util import TimeUtils
 from engagement_database.data_models import HistoryEntryOrigin
+from storage.google_cloud import google_cloud_utils
 
 from src.engagement_db_coda_sync.sync_stats import CodaSyncEvents, EngagementDBToCodaSyncStats
 
 log = Logger(__name__)
 
 
+def get_coda_users_from_gcloud(dataset_users_file_url, google_cloud_credentials_file_path):
+    return json.loads(google_cloud_utils.download_blob_to_string(
+        google_cloud_credentials_file_path, dataset_users_file_url
+    ))
+
+
+def ensure_coda_datasets_up_to_date(coda, coda_config, google_cloud_credentials_file_path): 
+    """
+    Ensures coda datasets are up to date based on coda configuration. 
+
+    :param coda: Coda instance to add the message to.
+    :type coda: coda_v2_python_client.firebase_client_wrapper.CodaV2Client
+    :param coda_config: Coda sync configuration.
+    :type coda_config: src.engagement_db_coda_sync.configuration.CodaSyncConfiguration
+    :param google_cloud_credentials_file_path: Path to a Google Cloud service account credentials file 
+                                               to use to access the credentials bucket.
+    :type google_cloud_credentials_file_path: str
+    """
+    all_datasets_have_user_file_url = all(
+        dataset_config.dataset_users_file_url is not None for dataset_config in coda_config.dataset_configurations)
+
+    default_config_user_ids = []
+    if not all_datasets_have_user_file_url:
+        assert coda_config.project_users_file_url is not None, \
+         f"Specify user ids for coda datasets in CodaDatasetConfiguration or user ids for this project in CodaSyncConfiguration"
+        default_config_user_ids = get_coda_users_from_gcloud(coda_config.project_users_file_url, google_cloud_credentials_file_path)
+
+    ws_correct_dataset_code_scheme = coda_config.ws_correct_dataset_code_scheme
+    for dataset_config in coda_config.dataset_configurations:
+        config_user_ids = []
+        if dataset_config.dataset_users_file_url:
+            config_user_ids = get_coda_users_from_gcloud(dataset_config.dataset_users_file_url, google_cloud_credentials_file_path)
+        else:
+            config_user_ids = default_config_user_ids
+
+        coda_user_ids = coda.get_dataset_user_ids(dataset_config.coda_dataset_id)
+        if coda_user_ids:
+            config_user_ids = list(set(config_user_ids) - set(coda_user_ids))
+            coda.update_dataset_user_ids(dataset_config.coda_dataset_id, config_user_ids)
```

Comment:
I think this should be reworked so it's a set rather than an update (i.e. replace the user ids in Coda with the authoritative version referenced in this repository).

Reply:
updated.